go/felix/intdataplane/endpoint_mgr.go

```diff
@@ -173,70 +194,181 @@ func (m *endpointManager) OnUpdate(protoBufMsg interface{}) {
+	// state updates and mark endpoint statuses for update.
+	for ifaceName := range m.pendingIfaceUpdates {
+		log.WithField("ifaceName", ifaceName).Debug("(Re)marking iface as dirty.")
+		m.markEndpointStatusDirtyByIface(ifaceName)
+		// Clean up as we go...
 		delete(m.pendingIfaceUpdates, ifaceName)
 	}
 
-	if err := m.resolveWorkloadEndpoints(); err != nil {
-		log.WithError(err).Warn("Failed to resolve workload endpoints")
-		return err
+	// Now send any endpoint status updates.
+	m.updateEndpointStatuses()
+
+	return nil
+}
+
+func (m *endpointManager) markEndpointStatusDirtyByIface(ifaceName string) {
+	logCxt := log.WithField("ifaceName", ifaceName)
+	if epID, ok := m.activeWlIfaceNameToID[ifaceName]; ok {
+		logCxt.Info("Workload interface state changed; marking for status update.")
+		m.epIDsToUpdateStatus.Add(epID)
+	} else if epID, ok := m.activeIfaceNameToHostEpID[ifaceName]; ok {
+		logCxt.Info("Host interface state changed; marking for status update.")
+		m.epIDsToUpdateStatus.Add(epID)
+	} else {
+		// We don't know about this interface yet (or it's already been deleted).
+		// If the endpoint gets created, we'll do the update then. If it's been
+		// deleted, we've already cleaned it up.
+		logCxt.Debug("Ignoring interface state change for unknown interface.")
 	}
+}
 
-	if m.dirtyHostEndpoints {
-		if err := m.resolveHostEndpoints(); err != nil {
-			log.WithError(err).Warn("Failed to resolve host endpoints")
-			return err
+func (m *endpointManager) updateEndpointStatuses() {
+	log.WithField("dirtyEndpoints", m.epIDsToUpdateStatus).Debug("Reporting endpoint status.")
+	m.epIDsToUpdateStatus.Iter(func(item interface{}) error {
+		switch id := item.(type) {
+		case proto.WorkloadEndpointID:
+			status := m.calculateWorkloadEndpointStatus(id)
+			m.OnEndpointStatusUpdate(m.ipVersion, id, status)
+		case proto.HostEndpointID:
+			status := m.calculateHostEndpointStatus(id)
+			m.OnEndpointStatusUpdate(m.ipVersion, id, status)
 		}
-		m.dirtyHostEndpoints = false
+
+		return set.RemoveItem
+	})
+}
+
+func (m *endpointManager) calculateWorkloadEndpointStatus(id proto.WorkloadEndpointID) string {
+	logCxt := log.WithField("workloadEndpointID", id)
+	logCxt.Debug("Re-evaluating workload endpoint status")
+	var operUp, adminUp, failed bool
+	workload, known := m.activeWlEndpoints[id]
+	if known {
+		adminUp = workload.State == "active"
+		operUp = m.activeUpIfaces.Contains(workload.Name)
+		failed = m.wlIfaceNamesToReconfigure.Contains(workload.Name)
 	}
 
-	return nil
+	var status string
+	if known {
+		if failed {
+			status = "error"
+		} else if operUp && adminUp {
+			status = "up"
+		} else {
+			status = "down"
+		}
+	}
+	logCxt = logCxt.WithFields(log.Fields{
+		"known":   known,
+		"failed":  failed,
+		"operUp":  operUp,
+		"adminUp": adminUp,
+		"status":  status,
+	})
+	logCxt.Info("Re-evaluated workload endpoint status")
+	return status
```

Comment:
Looks like status will be "" if known is false.  Is that the intention?  (Probably worth a comment if so.)